feat: add informer metrics (#983)
defenseunicorns/pepr: docs/030_user-guide/100_metrics.md

Reviewer: Not a huge fan of the intermittent, random capitalization of some of our metrics (i.e. `pepr_Mutate` vs `pepr_cache_miss`).  Would look much more polished if we stuck with a single format.

Author: addressed in next push of PR

- `pepr_alerts`: A counter that increments when an alert event is triggered in the application.
- `pepr_Mutate`: A summary that provides the observed durations of mutation events in the application.
- `pepr_Validate`: A summary that provides the observed durations of validation events in the application.
- `pepr_cache_miss`: A gauge that provides the number of cache misses per window.

> - `pepr_resync_failure_count`: A gauge that provides the number of failures per resync operation.

Reviewer: What does "the number of failures per resync operation" mean, exactly?  Should people be scared of this?  If we assume a naive reading -- where "failures" == BAD -- then seeing a number in this metric might make an operator think there's something going wrong when that's not necessarily the case.

Is there a way to fluff this out a little more so it's not so intimidating?  Like, describing what "failure" means in a more approachable way, maybe?

Author: this is based on KFC [resyncFailureMax](https://github.com/defenseunicorns/kubernetes-fluent-client/blob/6a936fd89ed20de0f36c5d7ada96242be84cd92a/src/fluent/watch.ts#L457), we can but i don't know how. What is your suggestion?

Author: *resyncFailureCount meant to say

Reviewer: "Number of unsuccessful attempts before making the last successful watch connection."

It skips over the word "resync" -- so as not to define something in terms of itself -- and also tries to box-in a bit of what "failure" means -- unsuccessful connection attempts -- so that it's not so open-ended / upsetting.  I was thinking something like that but maybe it's still missing some nuance?

Author: sounds good

> - `pepr_Resync_Failure_Count`: A gauge that provides the number of failures per resync operation.

Author: ```suggestion
- `pepr_Resync_Failure_Count`: A gauge that provides the number of unsuccessful attempts at receiving an event within the last seen event limit before re-establishing a new connection.
```

## Environment Variables

> | `PEPR_MAX_CACHE_MISS_WINDOWS` | Amount of cache miss windows to emit  | default: `"INF"`  |

Reviewer: ```suggestion
| `PEPR_MAX_CACHE_MISS_WINDOWS` | Maximum number windows to emit `pepr_cache_miss` metrics for  | default: `"INF"`  |
```
Updated desc to include specific, full name of metric being manipulated to facilitate finding by people who're using search to sift through the docs.

Author: sounds good

Author: resolved in PR

  pepr_Validate{quantile="0.999"} 201.23339900001884
  pepr_Validate_sum 402.4275380000472
  pepr_Validate_count 2

  # HELP pepr_cache_miss Number of cache misses per window
  # TYPE pepr_cache_miss gauge
  pepr_cache_miss{window="2024-07-25T11:54:33.897Z"} 18
  pepr_cache_miss{window="2024-07-25T12:24:34.592Z"} 0
  pepr_cache_miss{window="2024-07-25T13:14:33.450Z"} 22
  pepr_cache_miss{window="2024-07-25T13:44:34.234Z"} 19
  pepr_cache_miss{window="2024-07-25T14:14:34.961Z"} 0

  # HELP pepr_retry_count Number of retries per count
  # TYPE pepr_retry_count gauge

>   pepr_retry_count{count="0"} 5

Reviewer: ```suggestion
  pepr_resync_failure_count{count="0"} 5
```
..?

Author: good eye

>   # HELP pepr_resync_Failure_count Number of retries per count
>   # TYPE pepr_resync_Failure_count gauge
>   pepr_resync_Failure_count{count="0"} 5
>   pepr_resync_Failure_count{count="1"} 4

Author: ```suggestion
  # HELP pepr_Resync_Failure_Count Number of retries per count
  # TYPE pepr_Resync_Failure_Count gauge
  pepr_Resync_Failure_count{count="0"} 5
  pepr_Resync_Failure_count{count="1"} 4
```
